### PR TITLE
Build/erc2981 compliant

### PR DESCRIPTION
**Pull request by pymachado**
The feat for [this issue](https://github.com/Fusyona/nft-marketplace/issues/53) was solved at the current [PR](https://github.com/Fusyona/nft-marketplace/pull/57)

### CODE REVIEW

File: test/marketplace/royalty-testUnits.ts [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+    let collectionContractWithoutERC2981: IERC721;
+    let collectionContractWithERC2981: FusyERC721CollectionWithRoyaltySupport;
+  
+    let indexOfOfferMapping: number;
+    let nftId = 1;
+    const nftPrice = ethers.utils.parseEther("10");
+    const newPrice = ethers.utils.parseEther("9");
+
+    async function getCollectionWithoutERC2981() {
+        await deployments.fixture(["MockERC721Collection"])
```
Comment (MakeMake23) [inline test/marketplace/royalty-testUnits.ts:41]: Inside each of these 3 functions you call `deployments.fixture()` for an individual contract. That's inefficient. It's better to put all the contracts you want to deploy inside the same fixture and call `deployments.fixture()` once before each test.

File: test/marketplace/royalty-testUnits.ts [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+        let mockCollectionWitoutERC2981 = await deployments.get("MockERC721Collection"); 
+        collectionContractWithoutERC2981 = await ethers.getContractAt("MockERC721Collection", mockCollectionWitoutERC2981.address);
```
Comment (MakeMake23) [inline test/marketplace/royalty-testUnits.ts:43]: These two lines can become one if you use `ethers.getContract()` instead.

File: contracts/Marketplace.sol [hunk 3]
```diff
@@ -477,7 +505,9 @@ contract Marketplace is IMarketplace, ERC1155Holder, Ownable, ERC721Holder {
 
         nft.listed = false;
         _transferRemainingToSender(offer.price + msg.value, counteroffer.price);
-        _payingBenefits(seller, counteroffer.price);
+        
+        uint256 royalties = _payRoyalties(counteroffer.collection, counteroffer.nftId, counteroffer.price);
+        _payingBenefits(seller, counteroffer.price, royalties);
```
Comment (MakeMake23) [inline contracts/Marketplace.sol:510]: Royalty payment here is not tested. Maybe you want to do it.

**Review by MakeMake23 — COMMENTED** [review]
Good job @pymachado. Check my comments. The rest is fine 👍

**Comment by MakeMake23** [comment]
@pymachado please review. I don't know why I can't set you as a reviewer. Maybe it's because you are the creator.

File: contracts/Marketplace.sol [hunk 4]
```diff
@@ -164,35 +164,57 @@ contract Marketplace is IMarketplace, ERC1155Holder, Ownable, ERC721Holder{
...
+            (address creator, uint256 royalty) = royaltyInfo(
+                collection,
+                tokenId,
+                salePrice
+            );
+            payable(creator).transfer(royalty);
+            emit RoyaltyPayment(collection, tokenId, creator, royalty);
+            return royalty;
+        }
+        return 0;
```
Comment (TheSnakeWitcher) [inline contracts/Marketplace.sol:202]: Maybe we could do a short-circuit here

```solidity
if ( !supportsRoyalties(collection) ) return 0 ;

(address creator, uint256 royalty) = royaltyInfo(
                collection,
                tokenId,
                salePrice
            );
payable(creator).transfer(royalty);
emit RoyaltyPayment(collection, tokenId, creator, royalty);
return royalty;
```

Comment (MakeMake23) [reply contracts/Marketplace.sol:202]: Roger that 👍

**Review by pymachado — COMMENTED** [review]
We are making a great job here @@MakeMake23.  Thank you @TheSnakeWitcher for your suggestions in our reviews.

**Comment by MakeMake23** [comment]
> We are making a great job here @@MakeMake23. Thank you @TheSnakeWitcher for your suggestions in our reviews.

Thank you all 🙌

File: test/marketplace/royalty-testUnits.ts [hunk 5]
```diff
@@ -46,52 +50,168 @@ describe("Royalty support tests", () => {
...
+    async function listNft(
+        collection: IERC721 | FusyERC721CollectionWithRoyaltySupport
+    ) {
         return await marketplace
             .connect(seller)
-            .list(collectionContractWithERC2981.address, nftId, nftPrice);
+            .list(collection.address, nftId, nftPrice);
     }
+
+    describe("buy()", () => {
```
Comment (TheSnakeWitcher) [inline test/marketplace/royalty-testUnits.ts:93]: It is equally clear and more concise to write `buy()` instead of `buy function` , nice

**Pull request merged**